flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBIncrementalCheckpointUtils.java

```diff
@@ -0,0 +1,80 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.contrib.streaming.state;
+
+import org.apache.flink.runtime.state.KeyGroupRange;
+
+import org.rocksdb.ColumnFamilyHandle;
+import org.rocksdb.RocksDB;
+import org.rocksdb.RocksDBException;
+
+import java.util.List;
+
+/**
+ * Utils for RocksDB Incremental Checkpoint.
+ */
+public class RocksDBIncrementalCheckpointUtils {
+
+	public static void clipDBWithKeyGroupRange(
```

Comment:
I wonder if clipping the database to avoid prefix check is an optimization or not? If we don't clip, the must seek the iterator and apply a single if to every key. This if is very predictable for the CPU because it always passed except for when we terminate the loop. This sounds rather cheap. What are your thoughts about why deleting ranges is the better approach?

Reply:
Ah, I don't see a clear benefit either, but I think it makes the loop code look cleaner to me. But If you think that we don't need to clip the database to avoid prefix check, that's also good to me and I will change it.

Reply:
I think the code will still look ok, it is just one more `if` (we even only need if in the cases where we would clip something). If this allows us to eliminate some amount of codes, test, move away from experimental features, and may be faster then I think it is a good idea.